openhands/server/session/manager.py

```diff
@@ -1,26 +1,126 @@
 
-from fastapi import WebSocket
+import socketio
 
 from openhands.core.config import AppConfig
 from openhands.core.logger import openhands_logger as logger
-from openhands.events.stream import session_exists
+from openhands.events.stream import EventStream, session_exists
 from openhands.runtime.base import RuntimeUnavailableError
 from openhands.server.session.conversation import Conversation
-from openhands.server.session.session import Session
+from openhands.server.session.session import ROOM_KEY, Session
 from openhands.storage.files import FileStore
+from openhands.utils.shutdown_listener import should_continue
+
+_REDIS_POLL_TIMEOUT = 1.5
+_CHECK_ALIVE_INTERVAL = 15
 
 
 @dataclass
 class SessionManager:
+    sio: socketio.AsyncServer
     config: AppConfig
     file_store: FileStore
+    local_sessions_by_sid: dict[str, Session] = field(default_factory=dict)
+    local_connection_id_to_session_id: dict[str, str] = field(default_factory=dict)
+    _last_alive_timestamps: dict[str, float] = field(default_factory=dict)
+    _redis_listen_task: asyncio.Task | None = None
+    _session_is_running_flags: dict[str, asyncio.Event] = field(default_factory=dict)
+    _has_remote_connections_flags: dict[str, asyncio.Event] = field(
+        default_factory=dict
+    )
 
-    def add_or_restart_session(self, sid: str, ws_conn: WebSocket) -> Session:
-        return Session(
-            sid=sid, file_store=self.file_store, ws=ws_conn, config=self.config
-        )
+    async def __aenter__(self):
+        redis_client = self._get_redis_client()
+        if redis_client:
+            self._redis_listen_task = asyncio.create_task(self._redis_subscribe())
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        if self._redis_listen_task:
+            self._redis_listen_task.cancel()
+            self._redis_listen_task = None
+
+    def _get_redis_client(self):
+        redis_client = getattr(self.sio.manager, 'redis', None)
+        return redis_client
+
+    async def _redis_subscribe(self):
+        """
+        We use a redis backchannel to send actions between server nodes
+        """
+        redis_client = self._get_redis_client()
+        pubsub = redis_client.pubsub()
+        await pubsub.subscribe('oh_event')
+        while should_continue():
+            try:
+                message = await pubsub.get_message(
+                    ignore_subscribe_messages=True, timeout=5
+                )
+                if message:
+                    await self._process_message(message)
+            except asyncio.CancelledError:
+                return
+            except:
+                try:
+                    asyncio.get_running_loop()
+                    logger.warning(
+                        'error_reading_from_redis', exc_info=True, stack_info=True
+                    )
+                except RuntimeError:
+                    return  # Loop has been shut down
```

Comment:
should we `os.Exit` here or anything?

Reply:
No - if there are any other graceful shutdown tasks that might interfere with those - ending the loop should be sufficient.